_posts/2019-04-11-integration-with-office365.md

```diff
@@ -0,0 +1,50 @@
+
+Well first of all I need to say that in order to make this integration happen (as well as other/future ones dedicated to emails sending via 3rd party APIs) I had to do some refactoring of an already existing integration with Google mail. Given the documentation from Microsoft and having existing Google integration I understood what it should look like, so I changed the shape of the code without changing its functionality. After that everything seemed to be ok, at least from a design point of view, but then came implementation issues…
+
+Microsoft Graph API library for ruby (v.1.0) is written in a generalized way: it has an xml file representing data structures & types and some code wrapping the response into ruby objects respecting types description. It’s very nice to have such a thin wrapper over data, but for me it was difficult to debug it. Ideally, if there was more documentation available or even a community forum it would have made the process a little easier.
+
+<div class="text-align: center">
+    <img src="/assets/images/office365/microsoft-graph.png" alt="Microsoft graph" class="img-responsive" /><br />
+</div>
+
+My biggest win (over my stupidity) was that I couldn’t use library types in calls to API, they only wrap the response. For instance I created a FileAttachment instance and tried to pass it to `post` method fairly assuming that it will be well serialized and sent to MS endpoint but instead I got some strange errors :) So my mistake was that I was trying to work with API in the terms and data structures it provided me with. Instead I had to send objects that I created and serialized myself, that didn’t seem comfortable to me, but finally it worked. 
+
+The main problem was with attachments: I could not attach files directly to my message, there’s a limitation of 4MB per POST request on the MS endpoint side, so we can only guess the size of a file to be uploaded, because POST request can contain headers info and their length may vary, so the final attachment file size is a bit less than 4 Mb. How much less -- you should calculate somehow by subtracting the headers length from 4MBs. It’s not very convenient, but it is what it is. The format of the attachment data that is to be sent should be passed as a Base64 encoded string. I expected the API library to take a regular ruby hash with parameters to be serialized and sent to the endpoint, but I wanted too much. Though I prepared attachment file data and converted it into base64-encoded string, it didn’t work because API library doesn’t do any transformations or serialization so I had to do the whole POST request 100% myself composing request body myself. It was a hash serialized to string that had file attachment base64-encoded data and some other params saying that this attachment is really a file attachment. 
+
+So the whole procedure took several steps:
+ 1. Create a draft message in outlook account
+ 2. Attach files to it with separate POST requests 
+ 3. Send the message
```

Comment:
@anianc 
For some reason `<ol>` and `<ul>` items don't show up properly (missing bullet icon/number) since we switched themes.

I don't know if this is because of some styling priority issue.
But I think we should either:

(1) fix this issue
(2) add numbers to the text to (hack) the solution

Notes:
- I have done the later (2 - hacky solution) since the theme switch so be careful to add this fix globally
- You can add style local to individual file to test/fix/not break other pages. See [example](https://raw.githubusercontent.com/OnePageCRM/OnePageCRM.github.io/master/_posts/2018-09-27-a-brief-history-of-time-zones.md).

Reply:
@CillianMyles  yes, there is the css reset rule in new theme which removes all list style types. 
I didn't remove it to not break the theme. 
I've added another one with this commit to overwrite the css for lists inside the blog area. 
All <ul> <ol> lists inside the posts should have their initial bullets or numbers. If you added some inline hacks they will overwrite the main css rules.
The question is, is it ok that all list will got their bullets now? like [here](https://developer.onepagecrm.com/blog/2018/04/26/code-ninja-2017-18-results/)  or [here](https://developer.onepagecrm.com/blog/2018/03/02/rails_to_aws_lambda_communication_and_authentication/) (I believe there were bullets before in old blog)?
Or it is better to add special class for list with bullet like .bulletlist. .numberlist. But then we will have to remember to add this class and I am not sure if we can add css classes with Markdown syntax  so we will have to use HTML. That is why I've decided for global change but just inside the posts areas.

Reply:
@CillianMyles  I've made a typo in css before. Now the bullets are visible. I've added small hack for two posts (the list were styled diferently in them) and removed one .orig file.

Reply:
Thanks @anianc - looks great!